Fix go scalar collection cast

lampajr opened the pull request:
Fixes https://github.com/microsoft/kiota/issues/4380

Given openapi:
```yaml
    post:
      description: 'Create post'
      requestBody:
        required: true
        content:
          application/json:
            schema:
              type: array
              items:
                format: int32
                type: integer
```

The expected `ToPostRequestInformation` would be:
```go
// ToPostRequestInformation create post
// returns a *RequestInformation when successful
func (m *PostsRequestBuilder) ToPostRequestInformation(ctx context.Context, body []int32, requestConfiguration *PostsRequestBuilderPostRequestConfiguration) (*i2ae4187f7daee263371cb1c977df639813ab50ffa529013b7437480d1ec0158f.RequestInformation, error) {
	requestInfo := i2ae4187f7daee263371cb1c977df639813ab50ffa529013b7437480d1ec0158f.NewRequestInformationWithMethodAndUrlTemplateAndPathParameters(i2ae4187f7daee263371cb1c977df639813ab50ffa529013b7437480d1ec0158f.POST, "{+baseurl}/posts", m.BaseRequestBuilder.PathParameters)
	if requestConfiguration != nil {
		requestInfo.Headers.AddAll(requestConfiguration.Headers)
		requestInfo.AddRequestOptions(requestConfiguration.Options)
	}
	requestInfo.Headers.TryAdd("Accept", "application/json")
	cast := make([]interface{}, len(body))
	for i, v := range body {
		cast[i] = v
	}
	requestInfo.SetContentFromScalarCollection(ctx, m.BaseRequestBuilder.RequestAdapter, "application/json", cast)
	return requestInfo, nil
}
```

Note the array _cast_:

```go
	cast := make([]interface{}, len(body))
	for i, v := range body {
		cast[i] = v
	}
```

Which is requried as you cannot pass `[]int32` as `[]interface{}` without this trick.

lampajr commented:
@microsoft-github-policy-service agree

lampajr commented:
I did not create a new test as there was already existing one checking that specific scenarion: `WritesRequestGeneratorBodyForScalarCollection`, the issue is that it was not expecting that _cast_ at all.

baywet reviewed (commented):
Great work! thanks for putting this together!
About dotnet: what does dotnet --version and dotnet --list-sdks return? what OS are you running things on?
Can you also add a entry to the changelog please?

lampajr commented:
> About dotnet: what does dotnet --version and dotnet --list-sdks return? what OS are you running things on?

Here all infos:
```
cat /etc/fedora-release
Fedora release 38 (Thirty Eight)

dotnet --version
8.0.102

dotnet --list-sdks
8.0.102 [/usr/lib64/dotnet/sdk]
```

> Can you also add a entry to the changelog please?

Updated https://github.com/microsoft/kiota/pull/4388/commits/7089517eaf4b63206ab756b7d9b0231f0430a474

lampajr commented:
> latest is 203, but because you're on linux, you might be getting only the 100 patch branch for some reason. any chance you could use a Codespace for the purpose of testing?

TBH I never used Codespace, I just tried but it looks like automatically setup a devo container with .NET 7 instead of .NET 8, maybe there are some configuration in the repo that are forcing that dev container setup (?)

baywet commented:
My bad, I forgot to upgrade the dev container configuration when we moved to net 8.
I just put together #4391 for future use.
Now, the CI is failing, which I expected since you couldn't run unit tests locally.
What I suggest you to do here it to:
1. create a temporary branch of this one in your repo
2. merge my changes from 4391 in that temp branch

This way you should be able to run/debug unit tests (I just tried in my own container instance).

Once you have fixes, you can cherry pick the changes from that temp branch onto the current branch.

(or if it's too much gitfu, you can wait for my other PR to get merged, and rebase your branch)

lampajr commented:
Thanks @baywet !

I was able to run test using ther devcontainer approach and now they should be fixed :heavy_check_mark: 

Is there a way to build the CLI from this branch (download it) and run against my example ? Just to be 100% sure the change works as expected and no regressions are introduced.

lampajr commented:
With commit b0337cfed64cf2dbcf30c879c669a52f3785fcab I fixed the broken test and I also added some additional checks on other test cases to check whether the cast is generated or not.

baywet reviewed (approved):
Thanks for the continuous work on this!

> Is there a way to build the CLI from this branch (download it) and run against my example ? Just to be 100% sure the change works as expected and no regressions are introduced.

yes, you can directly run `dotnet run -c Release --project src/kiota/kiota.csproj -- <arguments you would pass>` to test things out.
You can alternatively edit one of the launch.json profiles (make sure you don't commit it)
I'll hold merging until you give me a confirmation

lampajr commented:
>  yes, you can directly run `dotnet run -c Release --project src/kiota/kiota.csproj -- <arguments you would pass>` to test things out.

Thanks, it worked! Maybe that's something that could be added in [CONTRIBUTING.md](https://github.com/microsoft/kiota/blob/main/CONTRIBUTING.md) as this could be a very useful information, especially for those who are not familiar with `dotnet` ecosystem.

>  I'll hold merging until you give me a confirmation

I was able to run it against my use cases and it works fine! @baywet  green light fmpov :rocket:

baywet commented:
Thanks for the contribution!
Please go ahead a open another pr to the contributing file for the guidance we discussed about if you feel it could be valuable to others.
